app/presenters/tree_node_builder.rb

```diff
@@ -62,108 +69,126 @@ def build
                                                 "switch.png",
-                                                _("Switch: %{name}") % {:name => object.name})
-    when User                 then generic_node(object.name, "user.png")
-    when MiqSearch            then generic_node(object.description,
+                                                _("Switch: %{name}") % {:name => object.name}) },
+    "User"                       => -> { generic_node(object.name, "user.png") },
+    "MiqSearch"                  => -> { generic_node(object.description,
                                                 "filter.png",
-                                                _("Filter: %{filter_description}") % {:filter_description => object.description})
-    when MiqDialog            then generic_node(object.description, "miqdialog.png", object[0])
-    when MiqRegion            then miq_region_node
-    when MiqWidget            then generic_node(object.title, "#{object.content_type}_widget.png", object.title)
-    when MiqWidgetSet         then generic_node(object.name, "dashboard.png", object.name)
-    when Tenant               then generic_node(object.name,  "#{object.tenant? ? "tenant" : "project"}.png")
-    when VmdbTableEvm         then generic_node(object.name, "vmdbtableevm.png")
-    when VmdbIndex            then generic_node(object.name, "vmdbindex.png")
-    when Zone                 then zone_node
-    when Hash                 then hash_node
+                                                _("Filter: %{filter_description}") % {:filter_description => object.description}) },
+    "MiqDialog"                  => -> { generic_node(object.description, "miqdialog.png", object[0]) },
+    "MiqRegion"                  => -> { miq_region_node },
+    "MiqWidget"                  => -> { generic_node(object.title, "#{object.content_type}_widget.png", object.title) },
+    "MiqWidgetSet"               => -> { generic_node(object.name, "dashboard.png", object.name) },
+    "Tenant"                     => -> { generic_node(object.name, "#{object.tenant? ? "tenant" : "project"}.png") },
+    "VmdbTableEvm"               => -> { generic_node(object.name, "vmdbtableevm.png") },
+    "VmdbIndex"                  => -> { generic_node(object.name, "vmdbindex.png") },
+    "Zone"                       => -> { zone_node },
+    "Hash"                       => -> { hash_node },
+  }
+  # rubocop:enable all
+
+  build_node_hash.each do |key|
+    begin
+      new_key = "#{key}Decorator"
+      new_key.constantize # check that it is a valid class
+      build_node_hash[new_key] = build_node_hash[key]
+
+    # Is there are better way to do this?
+    rescue NameError # rubocop:disable Lint/HandleExceptions
     end
+  end
+
+  BUILD_NODE_HASH = build_node_hash.freeze
+
+  def build
+    # Execute the proc from the BUILD_NODE_HASH in the context of the instance
+    instance_exec(&BUILD_NODE_HASH[object.class.to_s])
```

Comment:
I'm thinking we should call this with base class.

At first I thought it would lookup class then base class, but now I'm thinking just base class.

```ruby
# possibly
proc = BUILD_NODE_HASH[object.class.to_s] || BUILD_NODE_HASH[object.class.base_class.to_s]
instance_exec(&proc)

# more probably
instance_exec(&BUILD_NODE_HASH[object.class.base_class.to_s])
```